internal/overlord/checkstate/checkers.go

```diff
@@ -0,0 +1,165 @@
+			url:     config.HTTP.URL,
+			headers: config.HTTP.Headers,
+		}
+
+	case config.TCP != nil:
+		return &tcpChecker{
+			name: config.Name,
+			host: config.TCP.Host,
+			port: config.TCP.Port,
+		}
+
+	case config.Exec != nil:
+		return &execChecker{
+			name:        config.Name,
+			command:     config.Exec.Command,
+			environment: config.Exec.Environment,
+			userID:      config.Exec.UserID,
+			user:        config.Exec.User,
+			groupID:     config.Exec.GroupID,
+			group:       config.Exec.Group,
+			workingDir:  config.Exec.WorkingDir,
+		}
+
+	default:
+		// This has already been checked when parsing the config.
+		panic("internal error: invalid check config")
+	}
+}
+
+type httpChecker struct {
+	name    string
+	url     string
+	headers map[string]string
+}
+
+func (c *httpChecker) check(ctx context.Context) error {
+	logger.Debugf("Check %q (HTTP): requesting %q", c.name, c.url)
+	client := &http.Client{}
+	request, err := http.NewRequestWithContext(ctx, "GET", c.url, nil)
+	for k, v := range c.headers {
+		request.Header.Set(k, v)
+	}
+
+	response, err := client.Do(request)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode < 200 || response.StatusCode > 299 {
+		return fmt.Errorf("received non-20x status code %d", response.StatusCode)
+	}
+	return nil
+}
+
+type tcpChecker struct {
+	name string
+	host string
+	port int
+}
+
+func (c *tcpChecker) check(ctx context.Context) error {
+	logger.Debugf("Check %q (TCP): opening port %d", c.name, c.port)
+
+	host := c.host
+	if host == "" {
+		host = "localhost"
+	}
+
+	var dialer net.Dialer
+	conn, err := dialer.DialContext(ctx, "tcp", net.JoinHostPort(host, strconv.Itoa(c.port)))
+	if err != nil {
+		return err
+	}
+	err = conn.Close()
+	if err != nil {
+		logger.Noticef("Check %q (TCP): unexpected error closing connection: %v", c.name, err)
+	}
+	return nil
+}
+
+type execChecker struct {
+	name        string
+	command     string
+	environment map[string]string
+	userID      *int
+	user        string
+	groupID     *int
+	group       string
+	workingDir  string
+}
+
+func (c *execChecker) check(ctx context.Context) error {
+	logger.Debugf("Check %q (Exec): running %q", c.name, c.command)
+
+	args, err := shlex.Split(c.command)
+	if err != nil {
+		return fmt.Errorf("cannot parse check command: %v", err)
+	}
+
+	cmd := exec.CommandContext(ctx, args[0], args[1:]...)
+	for k, v := range c.environment {
+		cmd.Env = append(cmd.Env, k+"="+v)
+	}
+	cmd.Dir = c.workingDir
+
+	// Start as another user if specified in the check config.
+	uid, gid, err := osutil.NormalizeUidGid(c.userID, c.groupID, c.user, c.group)
+	if err != nil {
+		return err
+	}
+	if uid != nil && gid != nil {
+		cmd.SysProcAttr.Credential = &syscall.Credential{
+			Uid: uint32(*uid),
+			Gid: uint32(*gid),
+		}
+	}
+
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		if ctx.Err() == context.DeadlineExceeded {
+			err = fmt.Errorf("exec check timed out")
```

Comment:
if the check timed out it should be SIGKILLed before returning from this method.

Or at least: SIGTERM, start 30second timer, SIGKILL

Reply:
I don't think there's any need for this: per `exec.CommandContext` [docs](https://pkg.go.dev/os/exec#CommandContext), the exec library already kills the process (SIGKILL) "if the context becomes done before the command completes on its own" (that would apply to cancellation or timeout).